crates/ntx-builder/src/state/mod.rs

```diff
@@ -146,13 +184,44 @@ impl State {
             }
 
             self.in_progress.insert(candidate);
-            return TransactionCandidate { account: account.latest_account(), notes }.into();
+            return TransactionCandidate {
+                account: account.latest_account(),
+                notes,
+                chain_tip: self.chain_tip.clone(),
+                chain_mmr: self.chain_mmr.clone(),
+            }
+            .into();
         }
         self.inject_telemetry();
 
         None
     }
 
+    /// The latest block number the state knows of.
+    pub fn chain_tip(&self) -> BlockNumber {
+        self.chain_tip.block_num()
+    }
+
+    /// Updates the chain tip and MMR block count.
+    ///
+    /// Blocks in the MMR are pruned if the block count exceeds the maximum.
+    fn update_chain_tip(&mut self, tip: BlockHeader) {
+        // Update MMR which lags by one block.
+        self.chain_mmr.add_block(self.chain_tip.clone(), true);
+        self.chain_mmr_block_count += 1;
+
+        // Set the new tip.
+        self.chain_tip = tip;
+
+        // Prune MMR if necessary.
+        if self.chain_mmr_block_count > Self::MAX_BLOCK_COUNT {
+            let pruned_block_height =
+                self.chain_mmr.chain_length().as_u32() - Self::PRUNED_BLOCK_COUNT;
+            self.chain_mmr.prune_to(..pruned_block_height.into());
+            self.chain_mmr_block_count = Self::PRUNED_BLOCK_COUNT;
+        }
+    }
```

Comment:
Unfortunately don't have a way of unit testing the logic in this PR. Depending on the constructors available for `PartialBlockchain` and `PartialMmr`, could look at factoring this logic out to be testable as a separate struct.